src/basegraph.h

```diff
@@ -0,0 +1,355 @@
+	/**
+	 * \brief add single input to a vertex, allocate data vertex if needed
+	 * @param opname operator to add input to
+	 * @param input name of data input
+	 * @return input vertex
+	 */
+	Vertex add_input(std::string opname, std::string inputs);
+
+	/**
+	 * \brief add outputs to an operator vertex, allocates outputs if needed
+	 * @param v vertex to add ouput to
+	 * @param outputs name of data outputs
+	 * @return vector of data vertices
+	 */
+	std::vector<Vertex> add_outputs(Vertex v, std::vector<std::string>& outputs);
+
+
+	/**
+	 * \brief add outputs to an operator , allocates outputs if needed
+	 * @param opname name of operator to add ouput to
+	 * @param outputs name of data outputs
+	 * @return vector of data vertices
+	 */
+	std::vector<Vertex> add_outputs(std::string opname, std::vector<std::string>& outputs);
+
+
+	/**
+	 * \brief add output to an operator vertex, allocates output if needed
+	 * @param v vertex to add ouput to
+	 * @param varname name of data output
+	 * @return data vertex
+	 */
+	Vertex add_output(Vertex v, std::string varname);
+
+	/**
+	 * \brief add output to an operator , allocates output if needed
+	 * @param opnameto add ouput to
+	 * @param varname name of data output
+	 * @return data vertex
+	 */
+	Vertex add_output(std::string opname, std::string varname);
+
+	/**
+	 * \brief set loss to be used at learning
+	 * @param name  loss name
+	 */
+	void set_loss(std::string name) {_loss= name;}
+
+	/**
+	 * topological_sort of vertices, used internally
+	 */
+	void sort_all();
+
+	/**
+	 * \brief gives a dot reprensention of the graph, for debgging purposes
+	 * @param out ostream to put text to
+	 */
+	void todot(std::ostream &out);
+
+	/**
+	 * \brief check, sort, do some computations on blob sizes
+	 */
+	void finalize();
+
+	/**
+	 * \brief set input data dimensions
+	 * @param inputdim
+	 */
+	void set_input_dim(std::vector<int>& inputdim);
+
+
+	/**
+	 * \brief set ouput  data name
+	 * @param output : data blob name
+	 */
+	void set_output_name(std::string output) {_outputname = output;}
+
+	/**
+	 * \brief get input vertices
+	 * @param Vertex from which inputs are needed
+	 * @return all inpit vertices
+	 */
+	std::vector<Vertex> inputs(Vertex);
+
+	/**
+	 * \brief get output vertices
+	 * @param Vertex from which outputs are needed
+	 * @return all output vertices
+	 */
+	std::vector<Vertex> outputs(Vertex);
+
+	/**
+	 * \brief get some precise dimension of some inpit of a vertex
+	 *
+	 * @param v vertex from which some input dimension is needed
+	 * @param num_input number of the input (generally 0 for first input)
+	 * @param ndim number of dimension needed
+	 *
+	 * @return the dimension
+	 */
+	int dim(Vertex v, int num_input, int ndim);
+
+	/**
+	 * \brief gives number of outputs of a vertex
+	 */
+	inline
+	int num_output(Vertex v) {return _graph[v].num_output;}
+
+	/**
+	 * \brief accessor for an operator vertex name
+	 */
+	inline
+	std::string opname(Vertex v) {return _graph[v].name;}
```

Comment:
Add `const`? (Applies to other functions as well)
```
inline std::string opname(Vertex v) const
```

Reply:
ok